### DISCUSSION TITLE
Disabling Movement

### DISCUSSION THREAD

**Discussion by Kayoz1984** [textadventures/quest#1291] [Creating games with Quest] [answered]
I have a player.move attribute with an integer.  I want to use this attribute for the player to move in between rooms, reducing the int. by 1 with each turn. When the int. reaches zero, I want to prompt the player to rest, replenishing the attibute with a new random number. 

I have all the code worked out so far, but I'm having difficulty with disabling my players ability to move between rooms when the player.move counter reaches zero. What would be a solution that would help me to achieve this in Quest?

**Comment by angelwedge** [comment]
EDIT: There's now a better way. See my reply below.

I can see a few ways to do this. If you just want to disable the `go` command (so that interacting with NPCs or similar can still move the player), I'll suggest a two options:

1. Create a new command with the same pattern as `go`, and put it in the room with the player. This command will be used in preference to the standard one, so you can make this extra command just display a message telling the player they need to rest. Using the default English language file, the pattern should be `go to (?<exit>.*)$|^go (?<exit>.*)$|^(?<exit>north|east|south|west|northeast|northwest|southeast|southwest|in|out|up|down|n|e|s|w|ne|nw|se|sw|o|u|d)$`.

You can disable movement when the move counter hits zero by doing `MoveObjectHere (fake_go)`, and let them move again by doing `AddToInventory (fake_go)`. (Commands in the current room override global ones; commands in the player's inventory have no effect).

2. Modify the existing `go` command, putting an `if` statement around it.
If you're using the web editor you can't easily change the core commands; but you could create a new command called `fake_go` or something, with the pattern (regular expression) `go to (?<exit>.*)$|^go (?<exit>.*)$|^(?<exit>north|east|south|west|northeast|northwest|southeast|southwest|in|out|up|down|n|e|s|w|ne|nw|se|sw|o|u|d)$`, and give it a script like:
```
if (player.move > 0) {
  do (go, "script", QuickParams ("exit", exit))
  player.move = player.move - 1
}
else {
  msg ("You're too tired to move")
}
```
This just calls the built-in `go` command after checking the attribute.

**Comment by DraigNET** [comment]
I've had a play around but it's been a long time since I've used ASL. I thought perhaps you could do something with this?

`<beforeenter type="script">
    if (player.stamina = 0) {
        msg ("You are out of stamina and must rest.")
        canmove = false
    }
    else {
        canmove = true
    }
</beforeenter>`
      
The script runs before entering a room. You can create a custom function for canmove, and use verbs for adding the directions.

**Reply by angelwedge** [reply]
If you did something like that, you would need to have it on every room. It makes more sense to put it on the player's `changedparent` script, which runs when they move anyway; or modify the `go` command.

However, it looks like there is in fact a core way to do this, which I think was added after the last time I had a detailed look at the `go` command.

**Reply by DraigNET** [reply]
That is indeed one of the issues I found using this method. I still couldn't get it work the way I wanted it to either way lol.

**Comment by DraigNET** [comment]
There really should be a simpler way of doing this. Having a built-in **canmove** would be ideal.

**Comment by angelwedge** [comment]
Before moving the player, the go command runs a function called `TestExitGlobal`, which tests if the player is able to move.

If you're on the web editor, you can't edit existing functions. So instead, you should set `player.notallowedtoexit` to a string you want to be displayed instead of moving them. You should set this when the player becomes unable to move (when they use the last point from the movement attribute), and be set back to `null` to let the player move again. The main problem with this is that if there are two different things that can stop the player moving, it becomes difficult to ensure they don't interfere with each other.

If you are using the desktop editor, you should be able to change this function. It takes one parameter, the exit the player is trying to go through, and returns `true` if they are allowed to go through it.
So you could modify it like this:
```
<function name="TestExitGlobal" parameters="exit" type="boolean">
    if (HasString(game.pov, "notallowedtoexit")) {
      msg(game.pov.notallowedtoexit)
      return (false)
    }
    else if (game.pov.move > 0) {
      return (true)
    }
    else {
      msg ("You are too tired to move. You should rest first.")
      return (false)
    }
  </function>
```
(In this case, I named the attribute `move` like it was in the question

**Reply by DraigNET** [reply]
Excellent advice

**Reply by angelwedge** [reply]
You could also use the same function to reduce the player's stamina/movement allowance, if you want to have all the code for this in the same place… but note that this function is still called for look-only exits, which probably don't take stamina to tell the player they can't go that way. So you'd make something like:
```
    if (HasString(game.pov, "notallowedtoexit")) {
      msg(game.pov.notallowedtoexit)
      return (false)
    }
    else if (exit.lookonly) {
      return (true)
    }
    else if (game.pov.move <= 0) {
      msg ("You are too tired to move. You should rest first.")
      return (false)
    }
    else {
      game.pov.move = game.pov.move - 1
      return (true)
    }
```

**Reply by KVonGit** [reply]
> Before moving the player, the `go` command runs a function called `TestExitGlobal`, which tests if the player is able to move.

It doesn't work for me.

<details><summary> Test Game </summary>

![image](https://github.com/user-attachments/assets/b18a4671-bd49-4aa3-aba2-f1c6e58124f1)

```xml
<!--Saved by Quest 5.8.6836.13983-->
<asl version="580">
  <include ref="English.aslx" />
  <include ref="Core.aslx" />
  <game name="Not Allowed to Exit">
    <gameid>fc390148-2ff9-4cd0-b822-89ceb7ab04ac</gameid>
    <version>1.0</version>
    <firstpublished>2024</firstpublished>
  </game>
  <object name="room">
    <isroom />
    <object name="player">
      <inherit name="editor_object" />
      <inherit name="editor_player" />
      <notallowedtoexit>You don't have a hall pass!</notallowedtoexit>
    </object>
    <object name="hall pass">
      <inherit name="editor_object" />
      <take />
      <ontake type="script">
        game.pov.notallowedtoexit = null
      </ontake>
    </object>
    <exit alias="north" to="hall">
      <inherit name="northdirection" />
    </exit>
  </object>
  <object name="hall">
    <inherit name="editor_room" />
    <exit alias="south" to="room">
      <inherit name="southdirection" />
    </exit>
  </object>
</asl>
```

</details>

**Reply by KVonGit** [reply]
[WorldModel/WorldModel/Core/CoreCommands.aslx#L402-L410](https://github.com/textadventures/quest/blob/5782f9f228b577aa52442543fcde1324d987e99d/WorldModel/WorldModel/Core/CoreCommands.aslx#L402-L410)

```cs
  <command name="go" pattern="[go]" unresolved="[UnresolvedLocation]">
    if (exit.visible) {
      if (exit.locked) {
        msg (exit.lockmessage)
      }
      else if (not TestExitGlobal(exit)) {
        // Do nothing, already handled
      }
      else if (exit.runscript) {
```

---
That is the only result when I search `TestExitGlobal`.

![image](https://github.com/user-attachments/assets/6fd55862-15df-45a9-aa90-af84f60b8a76)

**Reply by KVonGit** [reply]
https://github.com/textadventures/quest/blob/5782f9f228b577aa52442543fcde1324d987e99d/WorldModel/WorldModel/Core/CoreCommands.aslx#L437-L445

And `TestExitGlobal` has an `exit` parameter that isn't used, which doesn't hurt anything (I don't think), but it makes me wonder if Andy meant to come back and finish adding this functionality and forgot about it.

**Reply by KVonGit** [reply]
No wait... That should work in my test game. Or it should at least print the `game.pov.notallowedtoexit` string.

```
      else if (not TestExitGlobal(exit)) {
        // Do nothing, already handled
      }
```

**Reply by KVonGit** [reply]
Err... My Quest 5.8 is missing that function.

![image](https://github.com/user-attachments/assets/07dff040-b713-4994-bfb5-53e43b542a42)

---
![image](https://github.com/user-attachments/assets/47bc2629-d3c9-4cf8-b466-ce4c8a1b2fca)

---
Huh... That's not in the desktop version of Quest 5.8.

```xml
  <command name="go" pattern="[go]" unresolved="[UnresolvedLocation]">
    if (exit.visible) {
      if (exit.locked) {
        msg (exit.lockmessage)
      }
      else if (exit.runscript) {
        if (HasScript(exit, "script")) {
          do (exit, "script")
        }
      }
      else if (exit.lookonly) {
         msg ("[UnresolvedLocation]")
      } 
      else {
        if (HasString(exit, "message")) {
          if (not exit.message = "") {
            if (game.clearscreenonroomenter) {
              game.currentexitmessage = exit.message
            }
            else {
              msg(exit.message)
            }
          }
        }
        game.pov.parent = exit.to
      }   
    } 
    else {   
      msg ("[UnresolvedLocation]")
    }
  </command>
```

---
But it **is** in the online version. When I create this same test game in the web editor, it works.

![image](https://github.com/user-attachments/assets/df382c21-4603-4b8b-8425-841138b9d415)

**Reply by KVonGit** [reply]
WebEditor: `<!--Saved by Quest 5.8.9010.29474-->`

Desktop: `<!--Saved by Quest 5.8.6836.13983-->`

**Reply by KVonGit** [reply, marked as answer]
Here's a library with Angel's modified `TestExitGlobal` from [this post](https://github.com/textadventures/quest/discussions/1291#discussioncomment-11617184), which also includes the `go` command needed to work with this:

[GoPatch-Kayoz1984Mod.zip](https://github.com/user-attachments/files/18207481/GoPatch-Kayoz1984Mod.zip)

---
I also made a library/patch for general use, using the `TestExitGlobal` function and `go` command from the WebEditor app:

https://github.com/textadventures/quest/discussions/1292

**Reply by Kayoz1984** [reply]
Works like a charm! Thank you!

**Reply by KVonGit** [reply]
All thanks to @angelwedge! 😁  (I didn't even know that code was in Quest, haha.)

---
**EDIT**

Your question also uncovered an inconsistency. This will help out all the desktop people who want to use this in Quest 5.8. 👍 👍
